File: javascript/triangle/triangle.js
class Triangle {

    constructor(...sides) {
        this.sides = sides;
    }

    kind() {
        if (!this.isValid()) {
            throw new Error();
        }
        const size = new Set(this.sides).size;
        return ['equilateral', 'isosceles', 'scalene'][size - 1];
    }

    isValid() {
        const [a, b, c] = this.sides.sort((a, b) => a - b);
        return (
            this.sides.every(s => s > 0)
            && a + b >= c
        );
    }
}

module.exports = Triangle;

File: fsharp/triangle/Triangle.fs
module Triangle

open System

let isValid (triangle: float list): bool = 
    match List.sort triangle with
    | [a; b; c] ->
        a + b >= c && List.forall (fun side -> side > 0.0) triangle
    | _ -> false

let uniqueSides = Seq.distinct >> Seq.length

let equilateral (triangle: float list): bool = 
    isValid triangle && uniqueSides triangle |> (=) 1

let isosceles (triangle: float list): bool =
    isValid triangle && uniqueSides triangle |> (>=) 2

let scalene (triangle: float list): bool =
    isValid triangle && uniqueSides triangle |> (=) 3